File: scripts/build_skill_packages.py
#!/usr/bin/env python3
"""Build or verify the self-contained files shipped inside each agent skill."""

from __future__ import annotations

import argparse
import shutil
from pathlib import Path

SKILL_REFERENCES = {
    "use-knowledge-vault": ("protocol.md",),
    "init-knowledge-vault": ("protocol.md", "contract-schema.md"),
    "audit-knowledge-vault": ("protocol.md", "contract-schema.md"),
    "manage-current-focus": ("protocol.md",),
}

RUNNER = """#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.11"
# dependencies = ["PyYAML==6.0.3"]
# ///
from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from knowledge_loom.cli import main


if __name__ == "__main__":
    raise SystemExit(main())
"""


def expected_files(package_root: Path, skill_name: str) -> dict[Path, bytes]:
    expected = {
        Path("references") / name: (package_root / "references" / name).read_bytes()
        for name in SKILL_REFERENCES[skill_name]
    }
    expected[Path("scripts") / "knowledge-loom.py"] = RUNNER.encode()
    source_root = package_root / "src"
    for source in sorted((source_root / "knowledge_loom").glob("*.py")):
        expected[Path("scripts") / source.relative_to(source_root)] = source.read_bytes()
    return expected


def managed_files(skill_root: Path) -> set[Path]:
    files: set[Path] = set()
    for directory in ("references", "scripts"):
        root = skill_root / directory
        if root.exists():
            files.update(path.relative_to(skill_root) for path in root.rglob("*") if path.is_file())
    return files


def check_skill(package_root: Path, skill_name: str) -> list[str]:
    skill_root = package_root / "skills" / skill_name
    expected = expected_files(package_root, skill_name)
    errors: list[str] = []
    for relative, contents in expected.items():
        target = skill_root / relative
        if not target.is_file():
            errors.append(f"missing {target}")
        elif target.read_bytes() != contents:
            errors.append(f"stale {target}")
    for relative in sorted(managed_files(skill_root) - set(expected)):
        errors.append(f"unexpected {skill_root / relative}")
    return errors


def build_skill(package_root: Path, skill_name: str) -> None:
    skill_root = package_root / "skills" / skill_name
    for directory in ("references", "scripts"):
        target = skill_root / directory
        if target.exists():
            shutil.rmtree(target)
    for relative, contents in expected_files(package_root, skill_name).items():
        target = skill_root / relative
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_bytes(contents)
        if relative == Path("scripts") / "knowledge-loom.py":
            target.chmod(0o755)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--check", action="store_true", help="fail when generated packages are missing or stale")
    args = parser.parse_args()
    package_root = Path(__file__).resolve().parents[1]

    if args.check:
        errors = [
            error
            for skill_name in SKILL_REFERENCES
            for error in check_skill(package_root, skill_name)
        ]
        if errors:
            for error in errors:
                print(f"ERROR {error}")
            return 1
        print("PASS self-contained skill packages are current")
        return 0

    for skill_name in SKILL_REFERENCES:
        build_skill(package_root, skill_name)
        print(f"BUILT skills/{skill_name}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/build-skill-packages.mjs
#!/usr/bin/env node
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { build } from "esbuild";

export const PACKAGE_ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
export const SKILL_REFERENCES = {
  "use-knowledge-vault": ["protocol.md"],
  "init-knowledge-vault": ["protocol.md", "contract-schema.md"],
  "audit-knowledge-vault": ["protocol.md", "contract-schema.md"],
  "manage-current-focus": ["protocol.md"],
};

async function bundledRunner(packageRoot) {
  const result = await build({
    absWorkingDir: packageRoot,
    entryPoints: ["src/knowledge-loom/runner.mjs"],
    bundle: true,
    format: "esm",
    platform: "node",
    target: "node20",
    write: false,
    legalComments: "none",
    charset: "utf8",
    banner: {
      js: [
        "import { createRequire as __knowledgeLoomCreateRequire } from 'node:module';",
        "const require = __knowledgeLoomCreateRequire(import.meta.url);",
        "// Self-contained Knowledge Loom runner. Bundles yaml 2.9.0 (ISC); see ../licenses/yaml.txt.",
      ].join("\n"),
    },
  });
  return result.outputFiles[0].contents;
}

export async function expectedFiles(packageRoot, skillName) {
  const expected = new Map();
  for (const name of SKILL_REFERENCES[skillName]) {
    expected.set(path.join("references", name), fs.readFileSync(path.join(packageRoot, "references", name)));
  }
  expected.set(path.join("scripts", "knowledge-loom.mjs"), await bundledRunner(packageRoot));
  expected.set(path.join("licenses", "yaml.txt"), fs.readFileSync(path.join(packageRoot, "node_modules", "yaml", "LICENSE")));
  return expected;
}

export function managedFiles(skillRoot) {
  const files = new Set();
  const visit = (root) => {
    if (!fs.existsSync(root)) return;
    for (const entry of fs.readdirSync(root, { withFileTypes: true })) {
      const candidate = path.join(root, entry.name);
      if (entry.isDirectory()) visit(candidate);
      else if (entry.isFile()) files.add(path.relative(skillRoot, candidate));
    }
  };
  for (const directory of ["references", "scripts", "licenses"]) visit(path.join(skillRoot, directory));
  return files;
}

export async function checkSkill(packageRoot, skillName) {
  const skillRoot = path.join(packageRoot, "skills", skillName);
  const expected = await expectedFiles(packageRoot, skillName);
  const errors = [];
  for (const [relative, contents] of expected) {
    const target = path.join(skillRoot, relative);
    if (!fs.statSync(target, { throwIfNoEntry: false })?.isFile()) errors.push(`missing ${target}`);
    else if (!fs.readFileSync(target).equals(contents)) errors.push(`stale ${target}`);
  }
  for (const relative of [...managedFiles(skillRoot)].sort()) {
    if (!expected.has(relative)) errors.push(`unexpected ${path.join(skillRoot, relative)}`);
  }
  return errors;
}

export async function buildSkill(packageRoot, skillName) {
  const skillRoot = path.join(packageRoot, "skills", skillName);
  for (const directory of ["references", "scripts", "licenses"]) {
    fs.rmSync(path.join(skillRoot, directory), { recursive: true, force: true });
  }
  for (const [relative, contents] of await expectedFiles(packageRoot, skillName)) {
    const target = path.join(skillRoot, relative);
    fs.mkdirSync(path.dirname(target), { recursive: true });
    fs.writeFileSync(target, contents);
    if (relative === path.join("scripts", "knowledge-loom.mjs")) fs.chmodSync(target, 0o755);
  }
}

export async function main(arguments_ = process.argv.slice(2)) {
  const check = arguments_.includes("--check");
  if (arguments_.some((argument) => argument !== "--check")) throw new Error(`unknown argument: ${arguments_.find((argument) => argument !== "--check")}`);
  if (check) {
    const errors = [];
    for (const skillName of Object.keys(SKILL_REFERENCES)) errors.push(...await checkSkill(PACKAGE_ROOT, skillName));
    if (errors.length) {
      for (const error of errors) console.error(`ERROR ${error}`);
      return 1;
    }
    console.log("PASS self-contained skill packages are current");
    return 0;
  }
  for (const skillName of Object.keys(SKILL_REFERENCES)) {
    await buildSkill(PACKAGE_ROOT, skillName);
    console.log(`BUILT skills/${skillName}`);
  }
  return 0;
}

if (path.resolve(process.argv[1] ?? "") === fileURLToPath(import.meta.url)) {
  main().then((status) => { process.exitCode = status; }).catch((error) => {
    console.error(`ERROR ${error.message}`);
    process.exitCode = 1;
  });
}
